worker/provisioner/provisioner_test.go

```diff
@@ -585,6 +585,64 @@ func (s *ProvisionerSuite) TestPossibleTools(c *gc.C) {
 	)
 }
 
+var validCloudInitUserData = `packages:
+  - 'python-keystoneclient'
+  - 'python-glanceclient'
+preruncmd:
+  - mkdir /tmp/preruncmd
+  - mkdir /tmp/preruncmd2
+postruncmd:
+  - mkdir /tmp/postruncmd
+  - mkdir /tmp/postruncmd2
+package_upgrade: false
+`
+
+func (s *ProvisionerSuite) TestSetUpToStartMachine(c *gc.C) {
+	attrs := map[string]interface{}{
+		config.CloudInitUserDataKey: validCloudInitUserData,
+	}
+	newConfig, err := s.Environ.Config().Apply(attrs)
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.Environ.SetConfig(newConfig)
+	c.Assert(err, jc.ErrorIsNil)
+
+	task := s.newProvisionerTask(
+		c,
+		config.HarvestAll,
+		s.Environ,
+		s.provisioner,
+		&mockDistributionGroupFinder{},
+		mockToolsFinder{},
+	)
+	defer workertest.CleanKill(c, task)
+
+	// Create the machine and check the tools that get passed into StartInstance.
+	machine, err := s.BackingState.AddOneMachine(state.MachineTemplate{
+		Series: "quantal",
+		Jobs:   []state.MachineJob{state.JobHostUnits},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	result, err := s.provisioner.Machines(machine.MachineTag())
+	c.Assert(err, gc.IsNil)
+	c.Assert(result, gc.HasLen, 1)
+	c.Assert(result[0].Err, gc.IsNil)
+	apiMachine := result[0].Machine
+
+	v, err := apiMachine.ModelAgentVersion()
+	c.Assert(err, jc.ErrorIsNil)
+
+	startInstanceParams, err := provisioner.SetupToStartMachine(task, apiMachine, v)
+	c.Assert(err, jc.ErrorIsNil)
+	cloudInitUserData := startInstanceParams.InstanceConfig.CloudInitUserData
+	c.Assert(cloudInitUserData, gc.DeepEquals, map[string]interface{}{
```

Comment:
This appears to exactly match the sample cloudinit data above. Shouldn't there be Juju data here as well?

Reply:
It is the same data, and it shouldn't have juju stuff yet, the instanceConfig is used to create the config for the machine in the cloudcloud.Configure methods, it doesn't contain the rendered cloudinit.